src/shared/elf-util.c

```diff
@@ -614,6 +620,11 @@ static int parse_core(int fd, const char *executable, char **ret, sd_json_varian
         if (!c.dwfl)
                 return log_warning_errno(SYNTHETIC_ERRNO(EINVAL), "Could not parse core file, dwfl_begin() failed: %s", sym_dwfl_errmsg(sym_dwfl_errno()));
 
+#if HAVE_DWFL_SET_SYSROOT
+        if (sym_dwfl_set_sysroot && root && sym_dwfl_set_sysroot(c.dwfl, root) < 0)
+                return log_warning_errno(SYNTHETIC_ERRNO(EINVAL), "Could not set root directory, dwfl_set_sysroot() failed: %s", sym_dwfl_errmsg(sym_dwfl_errno()));
+#endif
```

Comment:
Please normalize root with `empty_or_root` (skip the call if we don't really need to change root), and fail otherwise if `sym_dwfl_set_sysroot` is not supported.

Reply:
> and fail otherwise if `sym_dwfl_set_sysroot` is not supported.

This contradicts what @bluca asked for, so what behavior we actually want?

Reply:
We should not fail here, as it's possible to build against a new version and run with an old version

Reply:
Also if root is non NULL have to use it, as it is our code that prepares it we don't need to do any normalization or check whether we actually need to use it.

Reply:
> Also if root is non NULL have to use it, as it is our code that prepares it we don't need to do any normalization or check whether we actually need to use it.

Please make our util lib generic. We normalize root param with empty_or_root everywhere.

Reply:
> We should not fail here, as it's possible to build against a new version and run with an old version

If root is different from /, we really should fail. Otherwise it would be extremely confusing?

Reply:
I meant not fail if sym_dwfl_set_sysroot is not supported, that should continue with best-effort, as a coredump is still useful to have, even without a decoded trace

Reply:
dlsym_many_or_warnv would bail out if the symbol cannot be found BTW. So this is not supported.

Reply:
ok I thought it would continue, then the function check can be dropped, but we should probably fix that separately

Reply:
@YHNdnzj so this is the version you suggest right,

```C
#if HAVE_DWFL_SET_SYSROOT
        if (!empty_or_root(root) && sym_dwfl_set_sysroot(c.dwfl, root) < 0)
                return log_warning_errno(SYNTHETIC_ERRNO(EINVAL), "Could not set root directory, dwfl_set_sysroot() failed: %s", sym_dwfl_errmsg(sym_dwfl_errno()));
#endif
```

Reply:
Yes. Additionally if set_sysroot is not supported let's not silently continue. I.e. something like:

```c
if (empty_or_root(root))
        root = NULL;

#if HAVE_DWFL_SET_SYSROOT
        if (root && sym_dwfl_set_sysroot(c.dwfl, root) < 0)
                return log_warning_errno(SYNTHETIC_ERRNO(EINVAL), "Could not set root directory, dwfl_set_sysroot() failed: %s", sym_dwfl_errmsg(sym_dwfl_errno()));
#else
        if (root)
                return log_warning_errno(SYNTHETIC_ERRNO(EOPNOTSUPP), ...);
#endif
```

Reply:
it should warn but continue if it's not supported, again this is ancillary functionality, the core functionality is collecting and storing the core file